### PR TITLE
Azure DevOps Webhooks for Batch Changes

### PR DESCRIPTION
**Pull request by varsanojidan**
Closes: https://github.com/sourcegraph/sourcegraph/issues/46735

📺 Loom: https://www.loom.com/share/136fff58c7ca4b148bd37006a31228ab (wat at 1.2x I talk slow)

This pull requests is a follow-up to https://github.com/sourcegraph/sourcegraph/pull/48391, it builds off of the infrastructure there to make use of the new webhook events we can now ingest from Azure DevOps.

Things I want feedback on:
1) ADO has 4 main review states: Approved, Approved with suggestions, Waiting for Author, and Rejected, I was not 100% certain how that should be represented so I made Approved -> Approved, and the other 3 -> Changes Requested. Let me know what you think this can be changed fairly easily.
2) ADO doesn't really have many webhook events, the ones associated with PRs are: `git.pullrequest.updated` and `git.pullrequest.merged`. The updated event is also pretty barebones, it doesn't really directly mention what in the PR was updated, and does not get triggered on pullrequest checks. Because of this, I sort of simulated my own events off of the `git.pullrequest.updated` event, parsing the message of the event in order to convert it to a more meaningful event type for Review state changes. Outside of changes to the review, any other `git.pullrequest.updated` event will cause a manual API fetch from ADO in order to update the Changeset.

 
## Test plan
Manual testing so far

If you want to test this, you can follow the setup instructions here: https://github.com/sourcegraph/sourcegraph/pull/47913

### CODE REVIEW

File: enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go [hunk 1]
```diff
@@ -42,6 +48,28 @@ func (h *AzureDevOpsWebhook) Register(router *fewebhooks.Router) {
 func (h *AzureDevOpsWebhook) handleEvent(ctx context.Context, db database.DB, codeHostURN extsvc.CodeHostBaseURL, event any) error {
 	ctx = actor.WithInternalActor(ctx)
 
+	// If the event is just a general PR update event, it is best to just not try to derive the state from it and
+	// pull it in manually 💪.
+	switch e := event.(type) {
+	case *azuredevops.PullRequestUpdatedEvent:
```
Comment (varsanojidan) [inline enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go:54]: Retrieving what actually changed from this event is quite cumbersome, so we just make a call to ADO

File: internal/extsvc/azuredevops/events.go [hunk 2]
```diff
@@ -28,6 +35,34 @@ func ParseWebhookEvent(eventKey AzureDevOpsEvent, payload []byte) (any, error) {
 	if err := json.Unmarshal(payload, target); err != nil {
 		return nil, err
 	}
+
+	// Azure DevOps doesn't give us much in the way of differentiating webhook events, so we are going
+	// to try to parse the event message so that we can ideally simulate the different event types.
+	if eventKey == PullRequestUpdatedEventType {
```
Comment (varsanojidan) [inline internal/extsvc/azuredevops/events.go:41]: This is where I simulate the ReviewState change events by parsing the message of the `git.pullrequest.update` event in order to decide which "review change event" it is.

**Comment by varsanojidan** [comment]
I gotta admit, normally I would have done quite a bit more testing before opening up a PR, but as it is about to be the weekend, and basically all of my reviewers will be enjoying their weekend by the time I wake up, I thought it would be best to open this now so that I can get some feedback.

Feel free to 🔥 me. tomorrow I'll be adding whatever testing I missed in order to get this up for review in time.

File: enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go [hunk 3]
```diff
@@ -42,6 +48,28 @@ func (h *AzureDevOpsWebhook) Register(router *fewebhooks.Router) {
 func (h *AzureDevOpsWebhook) handleEvent(ctx context.Context, db database.DB, codeHostURN extsvc.CodeHostBaseURL, event any) error {
 	ctx = actor.WithInternalActor(ctx)
 
+	// If the event is just a general PR update event, it is best to just not try to derive the state from it and
+	// pull it in manually 💪.
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go:52]: 💪

File: enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go [hunk 4]
```diff
@@ -42,6 +48,28 @@ func (h *AzureDevOpsWebhook) Register(router *fewebhooks.Router) {
 func (h *AzureDevOpsWebhook) handleEvent(ctx context.Context, db database.DB, codeHostURN extsvc.CodeHostBaseURL, event any) error {
 	ctx = actor.WithInternalActor(ctx)
 
+	// If the event is just a general PR update event, it is best to just not try to derive the state from it and
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go:51]: I would specify event types explicitly in this comment (`PullRequestUpdatedEvent`, `PullRequestMergedEvent`). I was confused for a minute trying to figure out whether `merged` event is also a `general PR update event` or was added in a switch statement by mistake :)

File: enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go [hunk 5]
```diff
@@ -82,3 +117,41 @@ func azureDevOpsPullRequestEventPRs(e azuredevops.PullRequestEvent) []PR {
 		},
 	}
 }
+
+// enqueueAzureDevOpsChangesetSyncFromEvent enqueues a sync request for a specified changeset in repo-updater.
+// This is used instead of deriving the changeset from the incoming webhook event when doing so
+// would be difficult.
+func (h *AzureDevOpsWebhook) enqueueAzureDevOpsChangesetSyncFromEvent(ctx context.Context, esID extsvc.CodeHostBaseURL, event azuredevops.PullRequestEvent) error {
+	// We need to get our changeset ID for this to work. To get _there_, we need
+	// the repo ID, and then we can use the merge request IID to match the
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go:126]:
```suggestion
	// the repo ID, and then we can use the merge request ID to match the
```

File: enterprise/internal/batches/state/state.go [hunk 6]
```diff
@@ -393,16 +390,16 @@ func combineCheckStates(states []btypes.ChangesetCheckState) btypes.ChangesetChe
 
 	switch {
 	case stateMap[btypes.ChangesetCheckStateUnknown]:
-		// If are pending, overall is Pending
+		// If there are unknown states, overall is Pending
```
Comment (sashaostrikov) [inline enterprise/internal/batches/state/state.go:393]: 🚨 Full stop police here, you know what to do 🚨

File: internal/extsvc/azuredevops/events.go [hunk 7]
```diff
@@ -3,20 +3,27 @@ package azuredevops
...
 )
 
 var (
-	PullRequestCreatedEventType AzureDevOpsEvent = "git.pullrequest.created"
-	PullRequestMergedEventType  AzureDevOpsEvent = "git.pullrequest.merged"
-	PullRequestUpdatedEventType AzureDevOpsEvent = "git.pullrequest.updated"
+	PULL_REQUEST_APPROVED_TEXT                  = "approved pull request"
+	PULL_REQUEST_APPROVED_WITH_SUGGESTIONS_TEXT = "has approved and left suggestions in pull request"
+	PULL_REQUEST_REJECTED_TEXT                  = "rejected pull request"
+	PULL_REQUEST_WAITING_FOR_AUTHOR_TEXT        = "is waiting for the author in pull request"
```
Comment (sashaostrikov) [inline internal/extsvc/azuredevops/events.go:14]: camel/PascalCase for the win (https://sourcegraph.slack.com/archives/C3B3SDBMY/p1677768847277869)
:)

(And also uniform with vars below)
```suggestion
	PullRequestApprovedText                  = "approved pull request"
	PullRequestApprovedWithSuggestionsText = "has approved and left suggestions in pull request"
	PullRequestRejectedText                  = "rejected pull request"
	PullRequestWaitingForAuthorText        = "is waiting for the author in pull request"
```

File: enterprise/internal/batches/types/changeset_event.go [hunk 8]
```diff
@@ -316,6 +355,18 @@ func (e *ChangesetEvent) Timestamp() time.Time {
 		t = ev.CommitStatus.CreatedOn
 	case *bitbucketcloud.RepoCommitStatusUpdatedEvent:
 		t = ev.CommitStatus.UpdatedOn
+	case *azuredevops.PullRequestUpdatedEvent:
+		t = ev.CreatedDate
```
Comment (sashaostrikov) [inline enterprise/internal/batches/types/changeset_event.go:359]: lol, still cannot accept that ADO doesn't have an update date

**Review by sashaostrikov — APPROVED** [review]
Nice work, awesome demo! :chef_kiss: :davai:
Left a couple of non-critical comments.

File: enterprise/internal/batches/types/changeset_event.go (hunk 8 above)
Comment (varsanojidan) [reply enterprise/internal/batches/types/changeset_event.go:359]: The CreatedDate here is actually when the event was fired so it does work here 😅

**Review by eseliger — APPROVED** [review]

**Review by mrnugget — APPROVED** [review]
Good job!

File: doc/admin/config/webhooks.md [hunk 9]
```diff
@@ -15,6 +15,7 @@ GitHub | 🟢 | 🟢 | 🟢
 GitLab | 🟢 | 🟢 | 🔴
 Bitbucket Server / Datacenter | 🟢 | 🟢 | 🔴
 Bitbucket Cloud | 🟢 | 🟢 | 🔴
+Azure DevOps | 🔴 | 🟢 | 🔴
```
Comment (courier-new) [inline doc/admin/config/webhooks.md:18]: Is this accurate? Why is batch changes still 🔴? 😅

Comment (varsanojidan) [reply doc/admin/config/webhooks.md:18]: You are absolutely right 🤦. Fixed [here](https://github.com/sourcegraph/sourcegraph/pull/48972).

**Pull request merged**